### ISSUE TITLE
[feat] Fan speed

### ISSUE THREAD

**Issue by Bergoglio** [swingerman/ha-dual-smart-thermostat#196]
Hi,

the fan option is a good news.
My question is : can You develop also a fan speed manager inside Dual-smart-thermostat ?
For example I have this setup : I have only fan-coils (two types, this [Aermec fix speeds](https://global.aermec.com/it/products/scheda-prodotto/?t=Terminali%20idronici&c=CAT_50HZ_UE&f=terminal&Code=FCZ) with 3 fixed fan speeds plus 0 and this [Aermec inverter](https://global.aermec.com/it/products/scheda-prodotto/?t=Terminali%20idronici&c=CAT_50HZ_UE&f=terminal&Code=FCZI) with variable fan speed from 0 to 100% that are both heater + cooler) ; I used on inverter type an ESP32 Chinese electronic board with ESPHome firmware and on fixed speeds type a Shelly relè.
It will be awesome if Dual-smart-thermostat can also calculate and publish the right fan speed.
My idea is :

1. first parameter  >  fan coil fixed speed or fan coil inverter
2. if fan coil fixed speed > parameter : max difference between target temperature and real temperature to choose speed 1, max difference between target temperature and real temperature to choose speed 2, etc. (max five speeds)
The logic is : in heating Dual-smart switch on toggle for heating so trough an automation is possible to give order to fan coil to open hot water valve (via relè, via modbus, etc.) and also Dual-smart publish the right speed (1, 2, 3 or so on) and trough an automation is possible to give order to fan coil to switch the right fan speed (via relè, via modbus, etc.) ; in cooling Dual-smart switch on toggle for cooling so trough an automation is possible to give order to fan coil to open cold water valve (via relè, via modbus, etc.) and also Dual-smart publish the right speed (1, 2, 3 or so on) and trough an automation is possible to give order to fan coil to switch the right fan speed (via relè, via modbus, etc.)
3. if fan coil inverter > parameter : minimum percentage for speed, max percentage for speed, difference between target temperature and real temperature to choose max percentage for speed
The logic is : in heating Dual-smart switch on toggle for heating so trough an automation is possible to give order to fan coil to open hot water valve (via relè, via modbus, etc.) and also Dual-smart publish the percentage for speed (a calculation, if I set that 20% min speed + 80% max speed + difference for max speed is 1 degree and the difference target / real is 0.6 the published percentage will be 56%) and trough an automation is possible to give order to fan coil to switch the right fan speed (via relè, via modbus, etc.); the same for cooling.
4. the maximum will be also if it is possible to define an external switch for quiet mode that limits the max speed > for example a "good night" switch that I manually activate for the night...

**Comment by swingerman** [comment]
At this point, we can turn on/off the switches from the thermostat.
I can create extra attributes for speed steps or percentages, which one can use in automation.
Regarding the quiet mode, extending the preset modes with fan speed should be the way to go. You will be able to choose a preset, like "sleep," and it will set everything for you.

**Comment by swingerman** [comment]
I would suggest making use of the built-in fan_mode support of the climate platform in home assistant:
https://developers.home-assistant.io/docs/core/entity/climate/#fan-modes

**Comment by Bergoglio** [comment]
I don't know what is the best solution (I'm not a developer)....
For now I made this solution : 
- Dual- smart controls two helper switches, one for heating and one for cooling
- an automation triggers when the heating switch change its state : if from off to on the automation executes the start-heating script, if from on to off the stop- heating script
- the start-heating script open hot water valve and execute a calculation target temp - real temp (from sensor), if the difference is less than 0,2 then it activates switch for first speed, if between 0,2 and 0,4 switch for second speed, more than 0,4 switch for third speed
- etc. etc.
It works, but it will be more simple if all the work about choose of speed is made by Dual-smart.

**Comment by swingerman** [comment]
Ok.
The logic to set proper fan speed can be tricky though.
In your setup you step up the speed by 0,2 degree difference.
So I will set up the logic around that and will make it configurable.

**Comment by Bergoglio** [comment]
Two elements :

1. I think that is better if in the yaml configuration there will be two parameters > number of speeds (some fan coil as 2 speeds + 0, others 3 speeds + 0, others 5 speeds + 0, etc. we cannot make a general rule) and degree difference step between speeds (like You told)
2. starting from my experience, I understood one limit of my automation : if the necessary speed change during a cycle nothing changes on the fancoil; example : night preset is on (20 C degree), real temperature goes to 19,9° C, Dual-smart switches on heating and the speed by my automation is 1, then Simple-scheduler changes preset to comfort (21° C) and the necessary speed is 3 but nothing happens and the fancoil remains in speed 1......
So, it is good if Dual-smart publishes the necessary speed and that it changes continuously the speed if there is an increase until the real temperature matches the target temperature.

**Comment by swingerman** [comment]
Just to clarify:

The thermostat will not control the switches but will provide the required speed stage and percentage.

Configuration options:

```yaml
hvac_speed_step: 0.2 # <- temperature step
hvac_speed_steps: 3 # <- fan stages
```

Note: I prefer not to use literal `fan speed` to avoid limiting the feature to fan devices.

Provided state attributes:

- hvac_speed_stage
- hvac_speed_percent

The speed per cent can be tricky. I will need to use the `hvac_speed_step` to calculate the percentage or maybe the difference from the tolerance. I'm not sure. If there is only a 1-degree difference, the percentage will be 100%, and if there are 2 degrees, it will still be 100%. What if the system will never reach the target temperature if it is below 10%?

**Comment by Bergoglio** [comment]
> Just to clarify:
The thermostat will not control the switches but will provide the required speed stage and percentage.

Yes, You are right !

By my point of view :
First configuration option
`hvac_fan_type: fixed or percentage`
so there will be
`hvac_fan_type: fixed`
`hvac_speed_step: 0.2 # <- temperature step`
`hvac_speed_steps: 3 # <- fan stages`
or
`hvac_fan_type: percentage`
`hvac_min_start: 20 # <- minim percentage for hardware to start the fan`
`hvac_max_stop: 80 # <- percentage for hardware to achieve max fan speed`
`hvac_start_percentage: 30 # <- percentage when starting`
`diff_max_speed: 1 # <- tolerance where the fan speed must be max`

I think the difference from the tolerance is the best way to calculate the percentage, this is the reason of `diff_max_speed:` : "at what difference the speed must be max ? At 1 degree, at 1.4, at 2, etc." .

> If there is only a 1-degree difference, the percentage will be 100%, and if there are 2 degrees, it will still be 100%

Obviusly I agree (and I like)

> What if the system will never reach the target temperature if it is below 10%?

We can define a minim percentage when the fan coil must start
The formula can be this :

A > calculated percentage
B > hvac_start_percentage:
C > percentage for every 0.1 difference
D > 0.1 steps of difference detected
E > hvac_max_stop:
F > hvac_min_start:
G > number of 0.1 steps from yaml setup
H > target temperature
I > detected temperature
L > diff_max_speed:

hvac_speed_percent = the greater between A and B only when is necessary to heat or cool, when heating and cooling are off the value is 0
where A is
A = the greater between E and C * D + F
C = (E - F) / G
D = (H - I) / 0.1 for heating and (I - H) / 0.1 for cooling
G = L / 0.1

**Comment by Bergoglio** [comment]
Hi,

how are You ?
Do You have any news, idea, doubt or plan about fan speed ?
I don't want to press, only I think that enhancement will be very usefull for others like for me.
I found ready-made boards that can control directly the fancoils, they are based on ESP32 chip and I'm working on the firmware ESPHome : the board will take speed from the thermostat, pay attention to water temperature inside the fancoil and then set the fan speed directly to the fancoil engine in 0-10v.
When all will be running, if You want, I can send You a video of the "case history" :) : a kitchen with heating/cooling fancoil, mvhr ventilation and cooker hood all working together......

**Comment by swingerman** [comment]
I need to finish up other features and I will focus on this one.

**Comment by swingerman** [comment]
I am implementing a more generic solution that doesn't depend on fan mode, and I will call it `hvac_power_level`.

**Comment by Bergoglio** [comment]
I don't understand at 100%.....
In this way (hvac_power) it will be fine for fancoils with static speeds ( something like this [Aermec 3 speeds](https://global.aermec.com/it/products/scheda-prodotto/?t=Terminali%20idronici&c=CAT_50HZ_UE&f=terminal&Code=FCZ) ) but in inverter fancoils ( something like this [Aermec inverter](https://global.aermec.com/it/products/scheda-prodotto/?t=Terminali%20idronici&c=CAT_50HZ_UE&f=terminal&Code=FCZI) ) You lose the advantage of an inverter fancoil > the possibility to range fan speed continuously

**Comment by swingerman** [comment]
I will calculate a percentage and a power level, which will be available at all times.

**Comment by swingerman** [comment]
There is no need to set which mode to use as both modes will be calculated and provided in the state of the thermostat.

**Comment by Bergoglio** [comment]
Ok, now I understand.
Good, I like it.

**Comment by swingerman** [comment]
Please check https://github.com/swingerman/ha-dual-smart-thermostat/releases/tag/v0.9.8.beta-3

Note: the feature is not covered with test yet, so there can be bugs

It uses fallback values fo everything
defaults:
hvac_power_levels: 5
hvac_power_min: 1
hvac_power_max: 5
hvac_power_tolerance: 1 or 33 (based on the metric system of HA)

To enable you need to set either `hvac_power_levels` or `hvac_power_tolerance`

`hvac_power_tolerance` sets the sensitivity. The maximum difference between the current temp and the target temp.
In other words the levels and percentages will be calculated inside this tolerance.

**Comment by Bergoglio** [comment]
Ok, quite late I start tests....
One question out of this item : I saw the function "sensor_stale_duration", but how must I write the time in yaml code ?

**Comment by swingerman** [comment]
hh:mm:ss

**Comment by Bergoglio** [comment]
Ok, thanks

**Comment by Jerroder** [comment]
Hi @swingerman,
Sorry to update this a year later but I don't understand how to use this feature.

> I would suggest making use of the built-in fan_mode support of the climate platform in home assistant:
https://developers.home-assistant.io/docs/core/entity/climate/#fan-modes

This is exactly what I'm looking for but it's not supported by `platform: dual_smart_thermostat`.

> hvac_power_levels: 5
hvac_power_min: 1
hvac_power_max: 5

When I set this, I still don't have speed control. I was expecting something like this https://www.home-assistant.io/dashboards/features/#climate-fan-modes
<img width="252" height="126" alt="Image" src="https://github.com/user-attachments/assets/fe8e99ca-133c-465e-8797-ef104430b0c0"/>

Have I misunderstood?

**Comment by Bergoglio** [comment]
Hi @Jerroder ,

Dual Smart Thermostat works in this way :

1. You must set these parameters in yaml
    hvac_power_levels:    #number of fancoil's speeds 
    hvac_power_min:       #the first speed (normally 1, the lowest)
    hvac_power_max:      #the last speed (depends by fancoil, it is the highest)
    hvac_power_tolerance:      #the difference in celsius degree where must in use the max speed
2. in developer instruments > states You can find hvac_power_level:  (if You want to use a fixed speeds fancoil) and hvac_power_percent (if You want to use an inverter fancoil)
3. so, You must extract one of the two parameters in point 2 and use it in automations.

**Comment by Jerroder** [comment]
Hi @Bergoglio, thank you for your answer!

So as I understand it, it's not possible to set it manually? It's done automatically using `hvac_power_tolerance` but I can't set `hvac_power_levels` to 4 by pressing a button?

**Comment by Bergoglio** [comment]
> Hi [@Bergoglio](https://github.com/Bergoglio), thank you for your answer!
> 
> So as I understand it, it's not possible to set it manually? It's done automatically using `hvac_power_tolerance` but I can't set `hvac_power_levels` to 4 by pressing a button?

Exactly
